app_dart/lib/src/model/firestore/ci_staging.dart

```diff
@@ -0,0 +1,157 @@
+  int? get total => int.tryParse(fields![kTotalField]!.integerValue ?? '');
+
+  /// Mark a [checkRun] for a given [stage] with [conclusion].
+  ///
+  /// Returns a [StagingConclusion] record or throws. If the check_run was
+  /// both valid and recorded successfully, the record's `remaining` value
+  /// signals how many more tests are running. Returns the record (valid: false)
+  /// otherwise.
+  static Future<StagingConclusion> markConclusion({
+    required FirestoreService firestoreService,
+    // required String documentNameString,
+    required RepositorySlug slug,
+    required String sha,
+    required String stage,
+    required String checkRun,
+    required String conclusion,
+  }) async {
+    final logCrumb = 'markConclusion(${slug.owner}_${slug.name}_${sha}_$stage, $checkRun, $conclusion)';
+
+    // Marking needs to happen while in a transaction to ensure `remaining` is
+    // updated correctly. For that to happen correctly; we need to as perform a
+    // read of the document in the transaction as well. So start the transaction
+    // first thing.
+    final docRes = await firestoreService.documentResource();
+    final transactionResponse = await docRes.beginTransaction(
+      BeginTransactionRequest(
+        options: TransactionOptions(
+          readWrite: ReadWrite(),
+        ),
+      ),
+      kDatabase,
+    );
+    final transaction = transactionResponse.transaction;
+    if (transaction == null) {
+      log.warning('$logCrumb: transaction was null');
+      throw '$logCrumb: transaction was null';
+    }
+
+    var newRemaining = -1;
+
+    late final Document doc;
+
+    // transaction block
+    try {
+      // First: read the fields we want to change.
+      final documentPath = documentNameFor(sha: sha, slug: slug, stage: stage);
+      doc = await docRes.get(documentPath, mask_fieldPaths: [kRemainingField, checkRun], transaction: transaction);
+
+      // Fields and remaining _must_ be present.
+      if (doc.fields == null ||
+          doc.fields![kRemainingField] == null ||
+          doc.fields![kRemainingField]!.integerValue == null) {
+        log.warning('$logCrumb: missing field for $transaction / ${doc.fields}');
+        throw '$logCrumb missing fields';
+      }
+
+      final remaining = int.parse(doc.fields![kRemainingField]!.integerValue!);
+      newRemaining = remaining - 1;
+      // We will have check_runs scheduled after the engine was built successfully, so missing the checkRun field
+      // is an OK response to have. All fields should have been written at creation time.
+      if (doc.fields?[checkRun] == null || doc.fields?[checkRun]!.stringValue == null) {
+        log.info('$logCrumb: $checkRun not present in doc; remaining=$remaining');
+        await docRes.rollback(RollbackRequest(transaction: transaction), kDatabase);
+        return (valid: false, remaining: remaining);
+      }
+
+      // Now we can modify the document the change in the conculsion
```

Comment:
Something's wrong with this sentence. I think it's trying to say something like "Now we can set the conclusion in the document".

Reply:
I've changed some things around thanks to this comment.

1) I do want to record the new conclusion so it matches github calls. "valid = false" for the response
2) I only want to update remaining - 1 if the original conclusion is `kDefaultTaskStatus`. valid = true for the response.

I've updated the comment to reflect this.